mldatafind/find.py

```diff
@@ -35,61 +39,84 @@ def _data_generator(
     channels: Iterable[str],
     n_workers: int,
     thread: bool,
+    chunk_size: Optional[float] = None,
     **method_kwargs,
 ) -> Iterator:
 
-    memory_limit = MEMORY_LIMIT
-
     if thread:
         executor = ThreadPoolExecutor(n_workers)
     else:
         executor = ProcessPoolExecutor(n_workers)
 
+    logging.info(f"Finding {len(segments)} segments")
+
     with executor as exc:
         # keep track of current memory
         # and number of futures currently running
         current_memory = 0
-        futures = []
+        futures = {}
 
         # while there are still futures or segments to analyze
         while segments or futures:
 
             # submit jobs until memory limit is reached
-            while current_memory < memory_limit and segments:
+            while current_memory < MEMORY_LIMIT and segments:
                 segment = segments.pop()
 
                 duration = segment[1] - segment[0]
 
-                # TODO: memory depends on sample rate;
-                # for strain channels this is typically 16khz,
-                # but unknown for auxiliary channels
-                segment_memory = _calc_memory(len(channels), duration)
+                segment_memory = utils._estimate_memory(
+                    len(channels), duration
+                )
 
                 future = exc.submit(
                     method, channels, *segment, **method_kwargs
                 )
-                futures.append(future)
+                logging.info(
+                    f"Future submitted to query {duration} s of data"
+                    f"and {segment_memory:.2f} GB of memory"
+                )
+                futures[segment_memory] = future
                 current_memory += segment_memory
 
             # memory limit is saturated:
             # wait until any one future completes and yield
-            ready, futures = wait(futures, return_when=FIRST_COMPLETED)
-            print(futures, segments)
-            for future in ready:
-                yield future.result()
+            memories, done = utils.wait(futures)
+
+            for memory, f in zip(memories, done):
+                result = _handle_future(f)
+                yield result
+                current_memory -= memory
+
+
+def split_segments(segments: List, chunk_size: float):
```

Comment:
If you want to get rid of the numpy dependency here (since it's the only place it gets used), one way to refactor might be
```python
out = []
for segment in segments:
    start, stop = segment
    duration = stop - start
    if duration <= chunk_size:
        out.append(segment)
    else:
        num_chunks = (duration - 1) // chunk_size + 1
        for i in range(num_chunks):
            end = min(start + (i + 1) * chunk_size, stop)
            segment = Segment(start + i * chunk_size, end)
            out.append(segment)
    return out